### PR TITLE
refactor(flutter_bloc)!: remove package:provider

### LINKED ISSUE #712: [Discussion] Maintaining Provider Dependency
Hey everyone! 👋 

I am becoming concerned about the dependency on [provider](https://pub.dev/packages/provider) due to the increase in complexity. After each upgrade I find myself questioning whether it makes sense for the bloc library to continue to maintain the dependency on [provider](https://pub.dev/packages/provider) and wanted to start this discussion in order to see what everyone's thoughts are.

A few specific examples that come to mind include:

1. https://github.com/rrousselGit/provider/blob/00d0f8f8020dcfe876bd7826f176e1ca2feeaac7/packages/provider/lib/src/provider.dart#L288 - changes were made to prevent `Provider` from being used to provide `Streams`. Before bloc v1.0.0 `BlocProvider` was able to simply extend `Provider` which made the implementation extremely straightforward but after v1.0.0 (when bloc extended `Stream`) `BlocProvider` had to be refactored to extends `ValueDelegateWidget<T>` and implement `SingleChildCloneableWidget` with a manual `build` override to use an `InheritedProvider`. In this case, the implementation went from:

```dart
import 'package:flutter/widgets.dart';
import 'package:meta/meta.dart';
import 'package:provider/provider.dart';
import 'package:bloc/bloc.dart';

class BlocProvider<T extends Bloc<dynamic, dynamic>> extends Provider<T> {
  BlocProvider({
    Key key,
    @required ValueBuilder<T> builder,
    Widget child,
  }) : super(
          key: key,
          builder: builder,
          dispose: (_, bloc) => bloc?.dispose(),
          child: child,
        );

  BlocProvider.value({
    Key key,
    @required T value,
    Widget child,
  }) : super.value(
          key: key,
          value: value,
          child: child,
        );

  static T of<T extends Bloc<dynamic, dynamic>>(BuildContext context) {
    try {
      return Provider.of<T>(context, listen: false);
    } catch (_) {...}
  }
}
```

to

```dart
import 'package:flutter/widgets.dart';
import 'package:meta/meta.dart';
import 'package:provider/provider.dart';
import 'package:bloc/bloc.dart';

class BlocProvider<T extends Bloc<dynamic, dynamic>>
    extends ValueDelegateWidget<T> implements SingleChildCloneableWidget {
  final Widget child;

  BlocProvider({
    Key key,
    ValueBuilder<T> builder,
    Widget child,
  }) : this._(
          key: key,
          delegate: BuilderStateDelegate<T>(
            builder,
            dispose: (_, bloc) => bloc?.dispose(),
          ),
          child: child,
        );

  BlocProvider.value({
    Key key,
    @required T value,
    Widget child,
  }) : this._(
          key: key,
          delegate: SingleValueDelegate<T>(value),
          child: child,
        );

  BlocProvider._({
    Key key,
    @required ValueStateDelegate<T> delegate,
    this.child,
  }) : super(key: key, delegate: delegate);

  static T of<T extends Bloc<dynamic, dynamic>>(BuildContext context) {
    try {
      return Provider.of<T>(context, listen: false);
    } catch (_) { ... }
  }

  @override
  Widget build(BuildContext context) {
    return InheritedProvider<T>(
      value: delegate.value,
      child: Builder(
        builder: (context) => child,
      ),
    );
  }

  @override
  BlocProvider<T> cloneWithChild(Widget child) {
    return BlocProvider<T>._(
      key: key,
      delegate: delegate,
      child: child,
    );
  }
}
```

You might argue that we didn't add that much code but it wasn't immediately obvious to me what `ValueDelegateWidget` was for so I had to spent a significant amount of time understanding the implementation details of provider in order for me to feel comfortable refactoring the code.

2. `buildWithChild` introduced along with `SingleChildStatelessWidget`, etc... with the dependency on [nested](https://pub.dev/packages/nested) in `v4.0.0-dev`. While I like the idea, in practice I'm not happy with the implementation and there are several major limitations including not being able to use it as a `mixin` and also having to deal with a potentially `null` child when `buildWithChild` is called (unless I'm misunderstanding how to use it). I personally prefer the previous, simpler implementation because it was easy to understand and didn't involve custom `Widget` and `Element` implementations.

3. [flutter_bloc](https://pub.dev/packages/flutter_bloc) only requires a small subset of the functionality [provider](https://pub.dev/packages/provider) offers and I feel it would be much easier/less time-consuming to maintain the [flutter_bloc](https://pub.dev/packages/flutter_bloc) library if I didn't have to do a deep-dive of the [provider](https://pub.dev/packages/provider) implementation after every major change in order to understand how it will impact the bloc library.

The goal of this issue was just to start a discussion and see what everyone thinks. I'm more than happy to continue to maintain the dependency, but I just want to make sure that the cost of maintenance is justified by the benefit to the community. 

cc @rrousselGit, @brianegan, @larssn, @timrijckaert, @jorgecoca

### PR DESCRIPTION
**Pull request by felangel**
## Status

**READY**

## Breaking Changes

**YES**

### Explanation

This refactor technically is a breaking change because previously it was possible to use `Provider.of` and `context.read` with bloc/cubit instances. In addition, it was previously possible to import `package:provider/provider.dart` as a transitive dependency.

#### Migration

- `Provider.of<MyBloc>(context, listen: ...)` -> `BlocProvider.of<MyBloc>(context)`
- `context.read<MyBloc>()` -> `context.bloc<MyBloc>()`

If you were previously importing `package:provider/provider.dart` please ensure it is a direct dependency by adding it to the dependencies section of your `pubspec.yaml`

```yaml
flutter_bloc: ^6.0.6
provider: ^4.3.2+2 # make sure to add this if you directly import package:provider/provider.dart
```

## Description

- refactor(flutter_bloc): remove `package:provider` from `package:flutter_bloc` (closes #712)

### Feedback Wanted

You can give these changes a try by updating your `pubspec.yaml`

```yaml
  flutter_bloc:
    git:
      url: https://github.com/felangel/bloc
      ref: refactor/remove-provider-dependency
      path: packages/flutter_bloc
```

There should be no noticeable difference or breaking changes unless you are using `Provider.of` or `context.read` to interact with bloc/cubit instances.

If you experience any issues please leave a comment describing the problem, thanks! 🙏 

## Type of Change

- [ ] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [x] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [X] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: packages/flutter_bloc/lib/src/repository_provider.dart [hunk 1]
```diff
@@ -22,21 +26,15 @@ mixin RepositoryProviderSingleChildWidget on SingleChildWidget {}
...
-    Widget child,
-    bool lazy,
-  }) : super(
-          key: key,
-          create: create,
-          dispose: (_, __) {},
-          child: child,
-          lazy: lazy,
-        );
+    @required this.create,
```
Comment (alardizabal) [inline packages/flutter_bloc/lib/src/repository_provider.dart:34]: add assertion?

Comment (felangel) [reply packages/flutter_bloc/lib/src/repository_provider.dart:34]: Good catch!

**Review by jorgecoca — APPROVED** [review]
🚢  🇮🇹

**Comment by petermnt** [comment]
Hi @felangel,

We were using a `MultiProvider` which was using a list of `SingleChildWidget` as providers to provide both `Repository`'s en `Bloc`'s. 
This does not seem to be possible anymore. Am I missing an alternative or is this no longer supported by design?

**Pull request merged**